// szdt-admin-components/dist/esm/components/Empty/index.js
import "antd/es/card/style";
import _Card from "antd/es/card";
import "antd/es/empty/style";
import _Empty from "antd/es/empty";
import React from 'react';
import cls from 'classnames';
import "./index.less";
export default function Empty(props) {
  var _props$text = props.text,
    text = _props$text === void 0 ? '暂无数据' : _props$text,
    className = props.className,
    cardClassName = props.cardClassName,
    _props$img = props.img,
    img = _props$img === void 0 ? _Empty.PRESENTED_IMAGE_SIMPLE : _props$img,
    _props$show = props.show,
    show = _props$show === void 0 ? true : _props$show,
    _props$bordered = props.bordered,
    bordered = _props$bordered === void 0 ? true : _props$bordered,
    style = props.style;
  return show ? /*#__PURE__*/React.createElement(_Card, {
    bordered: bordered,
    className: cls('empty-content', cardClassName)
  }, /*#__PURE__*/React.createElement(_Empty, {
    image: img,
    description: text,
    className: className,
    style: style
  })) : null;
}
